### ISSUE TITLE
http3: RFC 9218 - Extensible Priority Scheme

### ISSUE THREAD

**Issue by chanh1964** [quic-go/quic-go#3470]
Hello developers.

Is there any work to implement RFC 9218 to quic-go?
I'm actually interested in (and in need of) such a work.

I guess I can give it a try if no work exists.
In that case, is there any hint on the files I should start with?

Thank you very much in advance.

**Comment by marten-seemann** [comment]
Sorry for the late response!

There's currently nobody working on implementing on RFC 9218, but such work would be highly appreciated!
Please comment on this thread if you're interested to submit a PR, then we can discuss the design.

**Comment by marten-seemann** [comment]
Chrome is planning to implement this feature: https://groups.google.com/a/chromium.org/g/blink-dev/c/eEeDzwtw5v0/m/wYqhn46XBQAJ

**Comment by SahibYar** [comment]
Hi @marten-seemann thank you for pointing to this in https://github.com/quic-go/quic-go/issues/3870

In reference to https://github.com/quic-go/quic-go/issues/437 "Dependency Weighing" mentioned in [RFC-7540](https://httpwg.org/specs/rfc7540.html#rfc.section.5.3.2)  is already deprecated as explain in RFC9218 in [Motivation for Replacing RFC 7540 Stream Priorities](https://datatracker.ietf.org/doc/html/rfc9218#name-motivation-for-replacing-rf) and [here](https://datatracker.ietf.org/doc/html/rfc9218#section-1-5)
>The design and implementation of RFC 7540 stream priority were observed to have shortcomings, as explained in Section 2. HTTP/2 [HTTP/2] has consequently deprecated the use of these stream priority signals. 

I have just started understanding `QUIC` and `quic-go` and want to implement Stream prioritization.

I was reading RFC9000 which [tells](https://datatracker.ietf.org/doc/html/rfc9000#section-3.1-4)
>Sending the first STREAM or STREAM_DATA_BLOCKED frame causes a sending part of a stream to enter the "Send" state. An implementation might choose to defer allocating a stream ID to a stream until it sends the first STREAM frame and enters this state, which can allow for better stream prioritization.
 
[and](https://datatracker.ietf.org/doc/html/rfc9000#section-13.3-4)
>Endpoints SHOULD prioritize retransmission of data over sending new data, unless priorities specified by the application indicate otherwise;


[and](https://datatracker.ietf.org/doc/html/rfc9000#section-2.3-3)
>A QUIC implementation SHOULD provide ways in which an application can indicate the relative priority of streams. An implementation uses information provided by the application to determine how to allocate resources to active streams.

Also we should try using strategy pattern, so that client can pass it's own priority algorithm as a parameter.  

## **Possible Solution I want to implement:**
As standard solution suggests, we can use [Priority Parameters in HTTP Headers](https://datatracker.ietf.org/doc/html/rfc9218#name-priority-parameters) like 
>priority = u=, I
Where u is urgence(int), and i(boolean) is incremental processing for _**signaling**_ request priority to origins ([and intermediaries](https://datatracker.ietf.org/doc/html/rfc9218#merging)). 

and for Reprioritization we can use [PRIORITY_UPDATE](https://datatracker.ietf.org/doc/html/rfc9218#section-7) frames. 
## **Questions in which I need help:**

1. For Reprioritization using `PRIORITY_UPDATE` do we have any **_control streams_**  as mentioned [here](https://datatracker.ietf.org/doc/html/rfc9218#section-7-2) ?
>PRIORITY_UPDATE frames are sent by clients on the control stream, allowing them to be sent independently of the stream that carries the response. This means they can be used to reprioritize a response or a push stream, or to signal the initial priority of a response instead of the Priority header field
2. Do we have any registry in quic-go (which can live on server, even if connection is closed) needed for implementing [4.3.1 Registration](https://datatracker.ietf.org/doc/html/rfc9218#name-registration)
3. Why do we need `streamQueue` in [framer.go](https://github.com/quic-go/quic-go/blob/master/framer.go#L32) ?
4. Since `quic-go` is maintaining 4 maps in [stream_maps.go](https://github.com/quic-go/quic-go/blob/master/streams_map.go#L57) do we need to prioritize streams among `outgoingBidiStreams`, `outgoingUniStreams`, `incomingBidiStreams`, `incomingUniStreams`
5. in streams_map.go why do we need `perspective protocol.Perspective` when we are assigning StreamID based on Server and Client ?
6. Please add any further detail which you think a new comer should know before starting to implement this feature.

**Comment by marten-seemann** [comment]
CloudFlare just published a blog post: https://blog.cloudflare.com/better-http-3-prioritization-for-a-faster-web/

**Comment by tobbee** [comment]
I'm interested in having priorities in QUIC stream for MOQ Transport via the [moqtransport](https://github.com/mengelbart/moqtransport) project, but it would of course be nice to have an API that also works for  H3. For MOQ, priorities are quite essential, starting with a control channel that should always have highest priority.

As references, I've looked at [moxygen](https://github.com/facebook/moxygen)+[mvfst](https://github.com/facebook/mvfst) for MOQ + QUIC  case for MOQ priorities, and [quiche](https://github.com/cloudflare/quiche) for H3 + RFC 9218. I've also looked at other implementation and finally used Claude Code with Opus 4 to analyze the different projects and come up with a plan for implementing this with a proper API in multiple steps. Below is the architecture description. 

I'm personally more interested in the MOQ part, but I hope this looks promising also for H3 and RFC 9218.

@marten-seemann  What are your thoughts on this?

# quic-go Priority Architecture Diagrams

## Overall Architecture

```
┌────────────────────────────────────────────────────────────────┐
│                        Application Layer                       │
│                                                                │
│  ┌─────────────────┐  ┌──────────────────┐  ┌───────────────┐  │
│  │  HTTP/3 Server  │  │  MOQ Application │  │ Custom Proto  │  │
│  │                 │  │                  │  │               │  │
│  │ Priority: u=2,i │  │ subPri=10        │  │ priority=high │  │
│  └────────┬────────┘  └────────┬─────────┘  └───────┬───────┘  │
│           │                    │                    │          │
└───────────┼────────────────────┼────────────────────┼───────-──┘
            │                    │                    │
            ▼                    ▼                    ▼
┌────────────────────────────────────────────────────────────────┐
│                      Priority API Layer                        │
│                                                                │
│  ┌─────────────────────────────────────────────────────────┐   │
│  │  stream.SetPriority(urgency, incremental)               │   │
│  │  stream.GetPriority() → Priority{Urgency: 2, Incr: true}│   │
│  └─────────────────────────────────────────────────────────┘   │
│                                                                │
│  ┌──────────────┐  ┌──────────────────┐  ┌────────────────┐    │
│  │ RFC9218      │  │ MOQPriority      │  │ CustomPriority │    │
│  │ Scheme       │  │ Scheme           │  │ Scheme         │    │
│  └──────────────┘  └──────────────────┘  └────────────────┘    │
└────────────────────────────────────────────────────────────────┘
                               │
                               ▼
┌────────────────────────────────────────────────────────────────┐
│                    Priority Scheduler                          │
│                                                                │
│  ┌─────────────────────────────────────────────────────────┐   │
│  │                   Urgency Level 0 (Highest)             │   │
│  │  ┌─────────────────┐        ┌─────────────────────┐     │   │
│  │  │ Sequential Heap │        │ Round-Robin Queue   │     │   │
│  │  │ [S1, S3, S7]    │        │ [S2]→[S4]→[S6]→⟲    │     │   │
│  │  └─────────────────┘        └─────────────────────┘     │   │
│  └─────────────────────────────────────────────────────────┘   │
│                                                                │
│  ┌─────────────────────────────────────────────────────────┐   │
│  │                   Urgency Level 1                       │   │
│  │  ┌─────────────────┐        ┌─────────────────────┐     │   │
│  │  │ Sequential Heap │        │ Round-Robin Queue   │     │   │
│  │  │ [S8, S11]       │        │ [S9]→[S10]→[S12]→⟲  │     │   │
│  │  └─────────────────┘        └─────────────────────┘     │   │
│  └─────────────────────────────────────────────────────────┘   │
│                                                                │
│                    ... Urgency Levels 2-7 ...                  │
└────────────────────────────────────────────────────────────────┘
                               │
                               ▼
┌────────────────────────────────────────────────────────────────┐
│                      QUIC Transport Layer                      │
│                                                                │
│  ┌─────────────────┐  ┌──────────────────┐  ┌──────────────┐   │
│  │ Packet Packer   │  │ Flow Controller  │  │ Congestion   │   │
│  │                 │  │                  │  │ Controller   │   │
│  │ GetNextStream() │  │ SendWindow: 64KB │  │ CWND: 10pkts │   │
│  └─────────────────┘  └──────────────────┘  └──────────────┘   │
└────────────────────────────────────────────────────────────────┘
```

## Priority Scheduler Detail

```
Priority Scheduler
├── getNextStream()
│   │
│   ├── [Urgency 0] ─── Check Sequential? ─── YES ──→ Return Stream
│   │                           │
│   │                          NO
│   │                           ▼
│   │                   Check Round-Robin? ─── YES ──→ Return Stream
│   │                           │
│   │                          NO
│   │                           ▼
│   ├── [Urgency 1] ─── Check Sequential? ─── YES ──→ Return Stream
│   │                           │
│   │                          NO
│   │                           ▼
│   │                   Check Round-Robin? ─── YES ──→ Return Stream
│   │                           │
│   │                          NO
│   │                           ▼
│   └── ... Continue through Urgency 7 ...
│
└── No streams with data → return nil
```

## Sequential vs Incremental Scheduling

### Sequential (Non-Incremental) Streams
```
Urgency 2 Sequential Heap:

         ┌─────┐
         │ S15 │  ← Top (highest priority)
         └──┬──┘
        ╱       ╲
    ┌──┴──┐   ┌──┴──┐
    │ S17 │   │ S23 │
    └─────┘   └─────┘

Scheduling: Always take S15 until it has no more data,
            then S17, then S23 (strict priority order)
```

### Incremental (Round-Robin) Streams
```
Urgency 2 Round-Robin Queue:

    ┌────┐     ┌────┐     ┌────┐     ┌────┐
 →  │S16 │ ──→ │S18 │ ──→ │S20 │ ──→ │S22 │ ↓
    └────┘     └────┘     └────┘     └────┘ │
       ↑                                    │
       └────────────────────────────────────┘
                    current ↓

Scheduling: S16 → S18 → S20 → S22 → S16 → ... (fair sharing)
            Each gets equal opportunity regardless of data amount
```

## MOQ Priority Encoding Example

```
MOQ Priority Encoding (64 bits):

┌────────┬──────────────┬──────────────┬─────────────────────┬─────────────────────┐
│ 6 bits │   8 bits     │   8 bits     │      21 bits        │      21 bits        │
├────────┼──────────────┼──────────────┼─────────────────────┼─────────────────────┤
│Reserved│ Subscriber   │ Publisher    │    Group ID         │    Object ID        │
│ (0)    │ Priority     │ Priority     │    (truncated)      │    (truncated)      │
└────────┴──────────────┴──────────────┴─────────────────────┴─────────────────────┘

Example Values:
- Control Stream:  0x0000000000000000 (all zeros = highest priority)
- Audio Stream:    0x0A0500000001A000 (subPri=10, pubPri=5, group=1, obj=0xA000)
- Video Stream:    0x140A00000002B000 (subPri=20, pubPri=10, group=2, obj=0xB000)

Comparison: Control > Audio > Video (lower value = higher priority)
```

## HTTP/3 Priority Flow

```
Client                          Server
  │                               │
  ├──── HTTP Request ────────────→│
  │     Headers:                  │
  │     :path: /video.mp4         │
  │     priority: u=2, i          │
  │                               │
  │                               ├── Parse Priority Header
  │                               ├── stream.SetPriority(2, true)
  │                               │
  │←──── HTTP Response ───────────┤
  │     Stream scheduled with     │
  │     urgency=2, incremental    │
  │                               │
  ├──── PRIORITY_UPDATE ─────────→│
  │     Stream: 4                 │
  │     Priority: u=1             ├── Update Stream Priority
  │                               ├── stream.SetPriority(1, false)
  │                               │
  │←──── Rescheduled Data ────────┤
  │     Now with urgency=1        │
  │                               │
```

## Memory Layout

```
StreamMeta Structure (64 bytes aligned):
┌─────────────────────┬────────┬─────────┐
│ StreamID (8)        │Priority│ Padding │
├─────────────────────┼────────┼─────────┤
│ HasData func (8)    │BytesPending func │
├─────────────────────┴──────────────────┤
│ LastServiced time.Time (16)            │
├────────────────────────────────────────┤
│ BytesServed (8)     │ HeapIndex (4)│pad│
├─────────────────────┴────────────-─┴───┤
│ QueueNode pointer (8)                  │
└────────────────────────────────────────┘

Optimized for cache line efficiency (64-byte boundary)
```

## Performance Characteristics

```
Operation Complexity:

├── AddStream         O(log n) for sequential, O(1) for incremental
├── RemoveStream      O(log n) for sequential, O(1) for incremental
├── UpdatePriority    O(log n) + remove + add
├── GetNextStream     O(1) best case, O(log n) worst case
└── RoundRobinAdvance O(1)

Memory Usage:
- Per stream: ~64 bytes metadata
- Per urgency level: 2 pointers (heap + round-robin)
- Total overhead: ~512 bytes + (64 * num_streams)

Cache Behavior:
- Hot path (GetNextStream) touches minimal memory
- Sequential heap benefits from cache locality
- Round-robin queue uses circular buffer for cache efficiency
```

## Integration Points

```
Modified Files:
quic-go/
├── config.go              (+10 lines: priority config)
├── connection.go          (+20 lines: scheduler init)
├── send_stream.go         (+30 lines: priority methods)
├── streams_map.go         (+15 lines: scheduler integration)
├── packet_packer.go       (+25 lines: priority scheduling)
│
├── internal/
│   └── priority/          (NEW: ~1500 lines)
│       ├── scheduler.go
│       ├── queue.go
│       ├── heap.go
│       └── round_robin.go
│
└── http3/
    ├── frames.go          (+50 lines: PRIORITY_UPDATE)
    ├── priority.go        (NEW: ~300 lines)
    └── server.go          (+40 lines: priority handling)

Total: ~2000 lines of new code, ~140 lines of modifications
```


- [mvfst HTTPPriorityQueue](https://github.com/facebook/mvfst): Reference implementation
- [moxygen MOQ priorities](https://github.com/facebook/moxygen): MOQ priority encoding
- [quiche h3 priorities](https://github.com/cloudflare/quiche): Production RFC 9218 implementation

**Comment by marten-seemann** [comment]
That proposal doesn't look totally crazy (contrary to the last [vibe-coded contribution](https://github.com/quic-go/quic-go/pull/5208)).

I'm a bit skeptical about the complexity estimates, and surprised that they're mentioned at all. `N` is fairly small and priority updates don't happen that often, so I don't think that's super relevant.

I'm even more skeptical about the LOC estimates, but I guess I'll have to see how the actual implementation looks like, this is hard to estimate.

---

That said, I've removed the "help wanted" label here, as this is a pretty complicated undertaking, requiring changes to both the QUIC and the HTTP/3 layer.

I believe that there would be great benefits to supporting priorities in the context of HTTP/3, Robin Marx has given some excellent talks on how page load times can be significantly improved when correctly prioritizing transfers.

If your company is using quic-go's HTTP/3 implementation and is interested in sponsoring work on this topic, please feel free to reach out to me via email.

**Comment by tobbee** [comment]
@marten-seemann Thanks for the response. I should probably have removed the LOC and complexity estimates, but I found them a bit amusing. The multi-level GetNextStream() complexity should be relevant since that is a new function. Using a priority queue is what I have seen in other implementations.

My own interest is in the basic prioritization layer to be used for MOQ where I think that moxygen/mvfst is a good reference. To the extent I will work on this, it will only be the basic prioritization and the MOQ mapping and not H3.

My company (Eyevinn Technology) does not use H3 at all, and MOQ is mainly an endeavor   driven by my own interest for streaming technologies, so I am unfortunately in no position to sponsor the project.

I have suggested quic-go as a foundation for a funded Open Source MoQ Relay project, but since our company is not a sponsor there either, I'm now out of the loop, so I cannot tell whether they will contact you.

I can understand that you don't want a huge blob of new code to review and take responsibility for in the future.

I intend to make a fork and see how far I get on implementing priorities for MOQ but trying to keep the basic priorities compatible with H3 in case it becomes good enough for other use than my own. 

If you start implementing priorities, some sort of pluggable architecture allowing for other priority schemes like the one defined in the MOQ spec would be very useful.

In any case, thanks for a great work.

**Comment by marten-seemann** [comment]
There's a new effort to implement RFC 9218 priorities for HTTP/2 in the Go standard library: https://github.com/golang/go/issues/75500
